rust/loqui_client/src/client.rs

```diff
@@ -3,56 +3,127 @@ use crate::waiter::ResponseWaiter;
 use crate::Config;
 use failure::Error;
 use futures::future::Future;
+use futures::sync::mpsc::{channel, Sender};
 use futures::sync::oneshot;
 use futures_timer::FutureExt;
-use loqui_connection::{convert_timeout_error, Connection, EncoderFactory, LoquiError};
+use loqui_connection::{Connection, LoquiError};
 use std::net::SocketAddr;
+use std::sync::atomic::{AtomicBool, Ordering::SeqCst};
+use std::sync::Arc;
+use std::sync::RwLock;
 use std::time::{Duration, Instant};
 use tokio::await;
-use tokio::net::TcpStream;
+use tokio::prelude::*;
 
-pub struct Client<F: EncoderFactory> {
-    connection: Connection<ConnectionHandler<F>>,
+pub struct Client {
+    connection: Connection<ConnectionHandler>,
     request_timeout: Duration,
+    handshake_deadline: Instant,
+    ready: Arc<AtomicBool>,
+    ready_waiter_tx: Sender<oneshot::Sender<()>>,
+    // TODO: might be able to set this unsafe
+    encoding: Arc<RwLock<Option<&'static str>>>,
```

Comment:
What do you folks think of doing an `unsafe` set of this? It's guarded by `self.encoding()`. For what it's worth, it doesn't seem to actually change the benchmarks.

Reply:
I wouldn't bother introducing an `unsafe` if it doesn't meaningfully improve benchmarks.